### PR TITLE
Nd jw user class

### LINKED ISSUE #3: User Class Favorite Recipes
As a user, I want to be able to add and remove favorite recipes that I find inside of the app.

- [ ] Create a method that will toggle favorite and unfavorite for a recipe

### LINKED ISSUE #4: User Class Favorite Recipes TDD
As a user, I want to be able to add and remove favorite recipes that I find inside of the app.

- [ ] Test the method that is created when a user toggles the favorite state for a recipe.
- [ ] Happy and sad path if possible.

### LINKED ISSUE #5: User Class Recipes To Cook
As a user, I want to be able to add recipes that I want to cook for a week to a list.

- [ ] Create a method that will store a recipe the user has toggled to cook for this week and return it to them.

### LINKED ISSUE #6: User Class Recipes To Cook TDD
As a user, I want to be able to add recipes that I want to cook for a week to a list.

- [ ] Create a method that will store a recipe the user has toggled to cook for this week and return it to them.
- [ ] Happy and sad path testing if possible.

### LINKED ISSUE #7: User Class Filter Recipes
As a user, I want to be able to filter my favorite recipes and my recipes to cook by type.

- [ ] Create a method that will return a user all of their favorite recipes or their recipes to cook for the week.

### LINKED ISSUE #8: User Class Filter Recipes TDD
As a user, I want to be able to filter my favorite recipes and my recipes to cook by type.

- [ ] Create a method that will return a user all of their favorite recipes or their recipes to cook for the week.
- [ ] Happy and sad path test if possible. (toUpperCase and toLowerCase)

### LINKED ISSUE #9: User Class Filter Saved Recipes
As a user, I want to be able to filter through my saved recipes by name or ingredient.

- [ ] Create a method that can filter a user's saved recipes by name and/or ingredient.

### LINKED ISSUE #10: User Class Filter Saved Recipes TDD
As a user, I want to be able to filter through my saved recipes by name or ingredient.

- [ ] Create a method that can filter a user's saved recipes by name and/or ingredient.
- [ ] Happy and sad path testing if applicable.

### LINKED ISSUE #53: Bug with Testing
- when running npm test, receive 2 errors that say: 
- Error in ./test/recipe-test.js
  Module not found: '../data/recipe-data' in '/Users/ashleymcdonald/turing/mod2projects/Whats-Cookin/test'

- Error in ./test/user-test.js
  Module not found: '../data/users-data' in '/Users/ashleymcdonald/turing/mod2projects/Whats-Cookin/test'

### LINKED ISSUE #55: Bug user-test finding a user
Found a bug in user-test.js where it is accessing a user incorrectly.

`TypeError: Cannot read property '0' of undefined`

### PR DESCRIPTION
**Pull request by npdarrington**
What’s this PR do?  
- This PR includes the user class as well as testing that user class with TDD.
- This PR allows a user to be created with an empty pantry list, empty favorite recipes list and empty recipes to cook list.
- A user will be able to save and remove recipes in their favorite recipes list.
- A user will be able to save and remove recipes in their recipes to cook list.
- A user will be able to filter recipes inside their favorites and recipes to cook lists by tag.
- A user will be able to search for a recipe by ingredient name and/or recipe name.

Where should the reviewer start?  
- Git pull this update.
- Open user.js, user-test.js files.

How should this be manually tested?  
- `npm test` inside of terminal.
- Check current code context in user.js class.

Any background context you want to provide?  
- No DOM updates for this, only data model and TDD.

What are the relevant tickets?  
- Closes #7 
- Closes #8 
- Closes #9 
- Closes #10 
- Closes #3 
- Closes #6 
- Closes #5 
- Closes #4 
- Closes #53 
- Closes #55  

Screenshots (if appropriate)  
- N/A

Questions: 
- N/A

### CODE REVIEW

File: src/user.js [hunk 1]
```diff
@@ -6,23 +6,26 @@ class User {
...
   }
+
   saveRecipe(recipe) {
     this.favoriteRecipes.push(recipe);
   }
 
-  removeRecipe(recipe) {
-    let i = this.favoriteRecipes.indexOf(recipe);
-    this.favoriteRecipes.splice(i, 1);
+  removeRecipe(recipeLocation, recipe) {
```
Comment (JWoltjen) [inline src/user.js:14]: I like this variable name a lot more

File: src/user.js [hunk 2]
```diff
@@ -6,23 +6,26 @@ class User {
...
+    this[recipeLocation].splice(recipeIndex, 1);
   }
 
   decideToCook(recipe) {
     this.recipesToCook.push(recipe);
   }
-  filterRecipes(type) {
-    return this.favoriteRecipes.filter(recipe => recipe.type.includes(type));
+
+  filterRecipes(recipeLocation, type) {
```
Comment (JWoltjen) [inline src/user.js:23]: added functionality here, as discussed

File: test/recipe-test.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { expect } from 'chai';
 
 import Recipe from '../src/recipe';
-import data from '../data/recipe-data';
+import data from '../src/data/recipe-data';
```
Comment (JWoltjen) [inline test/recipe-test.js:4]: Good catch here!

File: test/user-test.js [hunk 4]
```diff
@@ -1,61 +1,96 @@
...
   });
 
-  it('should be a function', function() {
+  it('should be a function', () => {
     expect(User).to.be.a('function');
   });
 
-  it('should initialize with an id', function() {
-    expect(user.id).to.eq(1);
+  it('should initialize with an id', () => {
```
Comment (JWoltjen) [inline test/user-test.js:24]: I like that we're using ES6 syntax. Let's make sure to change every one.

File: test/user-test.js [hunk 5]
```diff
@@ -1,61 +1,96 @@
...
+    expect(user.id).to.equal(1);
   });
 
-  it('should initialize with a name', function() {
-    expect(user.name).to.eq('Saige O\'Kon');
+  it('should initialize with a name', () => {
+    expect(user.name).to.equal('Saige O\'Kon');
```
Comment (JWoltjen) [inline test/user-test.js:29]: We decided that eq. is shorthand for equal, but that we should use equal instead.

File: test/user-test.js [hunk 6]
```diff
@@ -1,61 +1,96 @@
...
+
+  it('should be able to remove recipe from favorites', () => {
+    user.saveRecipe(recipe);
+    user.saveRecipe(recipe2);
+    expect(user.removeRecipe('favoriteRecipes', recipe)).to.deep.equal(user.favoriteRecipes[recipe2]);
+  });
+
+  it('should be able to remove recipe from recipes to cook', () => {
+    user.decideToCook(recipe);
+    user.decideToCook(recipe2);
```
Comment (JWoltjen) [inline test/user-test.js:93]: Love these additional tests.

**Review by JWoltjen — COMMENTED** [review]
Looks good to me Nathan, nice work!

**Review by aemcdonald — APPROVED** [review]

File: test/user-test.js [hunk 7]
```diff
@@ -1,61 +1,96 @@
 import { expect } from 'chai';
 
 import User from '../src/user';
-import data from '../data/users-data';
+import data from '../src/data/users-data';
 
-describe('User', function() {
+describe('User', () => {
```
Comment (hannakim91) [inline test/user-test.js:6]: Nice change to ES6 syntax!

File: test/user-test.js (hunk 4 above)
Comment (hannakim91) [reply test/user-test.js:24]: Agree - probably makes sense to make changes as we go and do a run-through at the end

File: test/recipe-test.js (hunk 3 above)
Comment (aemcdonald) [reply test/recipe-test.js:4]: nice catch!

File: src/user.js [hunk 8]
```diff
@@ -6,23 +6,26 @@ class User {
...
+
   saveRecipe(recipe) {
     this.favoriteRecipes.push(recipe);
   }
 
-  removeRecipe(recipe) {
-    let i = this.favoriteRecipes.indexOf(recipe);
-    this.favoriteRecipes.splice(i, 1);
+  removeRecipe(recipeLocation, recipe) {
+    let recipeIndex = this[recipeLocation].indexOf(recipe);
```
Comment (hannakim91) [inline src/user.js:15]: think this can be `const`

File: test/user-test.js (hunk 5 above)
Comment (aemcdonald) [reply test/user-test.js:29]: I prefer using equal because I think it improves readability

**Review by hannakim91 — APPROVED** [review]
Way more robust testing now! - just curious about the `filter` before passing `UserInfo` in to create a new `User`?

File: src/user.js (hunk 1 above)
Comment (aemcdonald) [reply src/user.js:14]: agree with Jeff, much better variable name here.

**Review by JWoltjen — APPROVED** [review]
Great work Nathan!~

**Pull request merged**